### PR TITLE
Make ApolloClient generic, add CallAdapter

### PR DESCRIPTION
**Pull request by digitalbuddha**
Part of #243 

- Make Apollo Client Generic based on `CallAdapter<T>` return type
- Create ApolloCallAdapter that implements CallAdapter<ApolloCall>
- Pass in CallAdapter<T> to Apollo Client Builder
- Pass`NewCall` through `CallAdapter`
- Clean up directory structure by creating impl folder (runtime was getting out of control), I'm open to other structures as well. Initial split was going to be /converter and /adapter but then some classes would need to be made public.

### CODE REVIEW

File: apollo-runtime/src/main/java/com/apollographql/android/impl/ApolloClient.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+          writer.value(customTypeAdapter.encode(value));
+        }
+      });
+      return this;
+    }
+
+    public ApolloClient<B> build() {
+      checkNotNull(okHttpClient, "okHttpClient is null");
+      checkNotNull(serverUrl, "serverUrl is null");
+      checkNotNull(callAdapter, "serverUrl is null");
```
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/android/impl/ApolloClient.java]: checkNotNull(callAdapter, "callAdapter is null");

**Comment by sav007** [comment]
Should we instead of `impl` package have `internal`?

**Comment by digitalbuddha** [comment]
The problem with internal is that the classes would still be public right?

**Comment by sav007** [comment]
No I meant to rename package `impl` to `internal` (like okHttp has)

File: apollo-runtime/src/main/java/com/apollographql/android/impl/ApolloClient.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+    return adapter.adapt(call);
+  }
+
+  public static class Builder<B> {
+    private OkHttpClient okHttpClient;
+    private HttpUrl serverUrl;
+    private final Map<ScalarType, CustomTypeAdapter> customTypeAdapters = new LinkedHashMap<>();
+    private Moshi.Builder moshiBuilder = new Moshi.Builder();
+    private CallAdapter<B> callAdapter;
```
Comment (BenSchwab) [inline apollo-runtime/src/main/java/com/apollographql/android/impl/ApolloClient.java]: Any reason to not default to the ApolloCallAdapter?

Comment (digitalbuddha) [inline apollo-runtime/src/main/java/com/apollographql/android/impl/ApolloClient.java]: I can't for the life of me figure out how to do that while keeping apollo client typesafe. Like then someone can do builder<Foo> and it'll bomb at runtime

File: apollo-runtime/src/main/java/com/apollographql/android/impl/ApolloClient.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+public final class ApolloClient<R> implements ApolloCall.Factory<R> {
+  public static <B> Builder<B> builder() {
+    return new Builder<>();
+  }
+
+  private final HttpUrl serverUrl;
+  private final Call.Factory httpCallFactory;
+  private final Map<ScalarType, CustomTypeAdapter> customTypeAdapters;
+  private final Moshi moshi;
+  private CallAdapter<R> adapter;
```
Comment (brianPlummer) [inline apollo-runtime/src/main/java/com/apollographql/android/impl/ApolloClient.java]: should be final

**Pull request merged**